Description:
 Fix integer overflow in IGMP protocol [SA-15:04] (CVE-2015-1414)
 .
 Updated patch from advisory revision 1.1
Origin: vendor, https://security.FreeBSD.org/patches/SA-15:04/igmp-errata.patch
Bug: https://www.freebsd.org/security/advisories/FreeBSD-SA-15:04.igmp.asc
Bug-Debian: https://bugs.debian.org/779195
Applied-Upstream: https://svnweb.freebsd.org/base?view=revision&revision=281232

--- a/sys/netinet/igmp.c
+++ b/sys/netinet/igmp.c
@@ -1533,8 +1533,7 @@
 		case IGMP_VERSION_3: {
 				struct igmpv3 *igmpv3;
 				uint16_t igmpv3len;
-				uint16_t srclen;
-				int nsrc;
+				uint16_t nsrc;
 
 				IGMPSTAT_INC(igps_rcv_v3_queries);
 				igmpv3 = (struct igmpv3 *)igmp;
@@ -1542,8 +1541,8 @@
 				 * Validate length based on source count.
 				 */
 				nsrc = ntohs(igmpv3->igmp_numsrc);
-				srclen = sizeof(struct in_addr) * nsrc;
-				if (nsrc * sizeof(in_addr_t) > srclen) {
+				if (nsrc * sizeof(in_addr_t) >
+				    UINT16_MAX - iphlen - IGMP_V3_QUERY_MINLEN) {
 					IGMPSTAT_INC(igps_rcv_tooshort);
 					return;
 				}
@@ -1552,7 +1551,7 @@
 				 * this scope.
 				 */
 				igmpv3len = iphlen + IGMP_V3_QUERY_MINLEN +
-				    srclen;
+				    sizeof(struct in_addr) * nsrc;
 				if ((m->m_flags & M_EXT ||
 				     m->m_len < igmpv3len) &&
 				    (m = m_pullup(m, igmpv3len)) == NULL) {
